lib/wm_util.ml

```diff
@@ -111,11 +111,71 @@ module MediaType = struct
     loop ranges
 end
 
-module Date = struct
-  open CalendarLib
+module type Date_sig = sig
+  type t
+  val now : unit -> t
+  val parse_rfc1123_date_exn : string -> t
+  val parse_rfc1123_date : string -> t option
+end
+
+module type PtimeNow_sig = sig
+  val now : unit -> Ptime.t
+end
+
+(** MakePtimeDate creates a Ptime based Date module, that can be used with the
+    Webmachine.Make(IO)(Date) functor.
+
+    Examples:
+    module DateMock = Wm_util.MakePtimeDate(struct
+      let now = fun () -> Ptime.v (17591, 0L)
+    end)
+
+    module DatePtimeClock = Wm_util.MakePtimeDate(struct
+      let now = Ptime_clock.now
+    end)
+
+    module DateMirage = Wm_util.MakePtimeDate(struct
+      let now = fun () -> Ptime.v @@ Pclock.now_d_ps clock
+    end)
+*)
+module MakePtimeDate(NOW:PtimeNow_sig) : Date_sig = struct
+
+  type t = Ptime.t
+
+  let now = NOW.now
 
   let parse_rfc1123_date_exn s =
-    Printer.Time.from_fstring "%a, %d %b %Y %H:%M:%S GMT" s
+    try Scanf.sscanf s "%3s, %d %s %4d %d:%d:%d %s" (
+      fun _wday mday mon year hour min sec tz ->
+        let months = [
+          "Jan", 1; "Feb", 2; "Mar", 3; "Apr", 4; "May", 5; "Jun", 6;
+          "Jul", 7; "Aug", 8; "Sep", 9; "Oct", 10; "Nov", 11; "Dec", 12
+        ] in
+        let parse_tz = function
+          | "" | "Z" | "GMT" | "UTC" | "UT" -> 0
+          | "PST" -> -480
+          | "MST" | "PDT" -> -420
+          | "CST" | "MDT" -> -360
+          | "EST" | "CDT" -> -300
+          | "EDT" -> -240
+          | s -> Scanf.sscanf s "%c%02d%_[:]%02d" (fun sign hour min ->
+              min + hour * (if sign = '-' then -60 else 60))
+        in
+        let mon = List.assoc mon months in
```

Comment:
Is the potential `Not_found` exception caught anywhere?  Or is it ok if it leaks from here?